foxpuppet/browser/notifications.py

```diff
@@ -0,0 +1,103 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from abc import ABCMeta
+
+from selenium.webdriver.common.by import By
+
+
+class BaseNotification(object):
+    """Abstract base class for any kind of notification."""
+
+    __metaclass__ = ABCMeta
+
+    def __init__(self, window, root):
+        self.window = window
+        self.root = root
+        self.selenium = self.window.selenium
+
+    @property
+    def label(self):
+        """Provide access to the notification label.
+
+        :returns: The notification label.
+        """
+        with self.selenium.context('chrome'):
+            return self.root.get_attribute('label')
+
+    @property
+    def origin(self):
+        """Provide access to the notification origin.
+
+        :returns: The notification origin.
+        """
+        with self.selenium.context('chrome'):
+            return self.root.get_attribute('origin')
+
+    def close(self):
+        """Close the notification."""
+        with self.selenium.context('chrome'):
+            self.root.find_anonymous_element_by_attribute(
+                'anonid', 'closebutton').click()
```

Comment:
The anonymous element functions are implemented as extension commands with custom endpoints on geckodriver that are separate and disparate from the normal element retrieval endpoints.

This is in part due to typing differences as the data input taken for `FirefoxWebElement.anonymous_children` and `FirefoxWebElement.find_anonymous_element_by_attribute` is incompatible with the data input types expected for `RemoteWebElement.find_element`/`RemoteWebElement.find_elements`. The anonymous element methods allow input such as null and dictionaries (with some very funky behaviour in how this is treated, if you read the Marionette implementation), which doesn’t correspond to the (de)serialisation types defined by the Selenium remote server or the webdriver-rust library. In other words, trying to send `{"using": "anon attribute", "value": {}}` will cause a type check failure (probably an illegal attribute error) because the value type is not recognised.

The object serialisation for the extension commands in Selenium is found here: https://github.com/SeleniumHQ/selenium/blob/master/py/selenium/webdriver/firefox/webelement.py
And the mapping of the strings passed to URL routing is done here: https://github.com/SeleniumHQ/selenium/blob/master/py/selenium/webdriver/firefox/remote_connection.py

But also because the commands provide non-standard behaviour that other drivers don’t supply, there is a requirement that they are published separately, under a vendor-prefixed namespace:

* `/session/$sessionId/moz/xbl/$id/anonymous_children`
* `/session/$sessionId/moz/xbl/$id/anonymous_by_attribute`

It would probably be possible to smooth over this difference in the client binding by overriding `FirefoxWebElement.find_element`/`FirefoxWebElement.find_elements`, look at the strategy and then call the appropriate command; whether that be the standard element retrieval commands or the anonymous element extensions. But again, that would require an API change in the upstream library, and I’m not sure that’s in scope for this patch.

Anyway, I hope this illuminates the situation (-: